common/variant/shredding.md

```diff
@@ -0,0 +1,244 @@
+
+The **untyped_value** is stored in a single binary column, rather than storing the value and metadata separately as is done in the unshredded binary format. The motivation for storing them separately for unshredded data is that this lets the engine encode and compress the metadata more efficiently when the fields are consistent across rows. We chose to combine them in the shredded fields: we expect the encoding/compression benefit to be lower, since in the case of uniform data, the values should be stored in typed columns. Using a single binary reduces the extra metadata required in the Parquet schema, which can be significant in some cases.
+
+# Data Skipping
+
+Shredded columns are expected to store statistics in the same format as a normal Parquet column. In general, the engine can only skip a row group or page if all rows in the **untyped_value** field are null, since it is possible for a `variant_get` expression to successfully cast a value from the **untyped_value** to the target type. For example, if **typed_value** is of type `int64`, then the string “123” might be contained in **untyped_value**, which would not be reflected in statistics, but could be retained by a filter like `where variant_get(col, “$.field”, “long”) = 123`. If **untyped_value** is all-null, then the engine can prune pages or row groups based on **typed_value**. This specification is not strict about what values may be stored in **untyped_value** rather than **typed_value**, so it is not safe to skip rows based on **typed_value** unless the corresponding **untyped_value** column is all-null, or the engine has specific knowledge of the behavior of the writer that produced the shredded data.
+
+# Shredding Semantics
+
+Variant defines a number of integer and decimal types of varying widths. When writing, it would be quite limiting to strictly enforce the mapping between Variant types and Parquet/Spark types. For example, if we chose to shred a field as `int64`, and encountered the value 123 encoded as `int32`, it seems preferable to write this to the **typed_value** column, even though it technically loses information about the type in the original Variant object, and would be reconstructed as an `int64`.
+
+On the other hand, storing arbitrarily casted values in the **typed_value** column could create inconsistent behavior before and after shredding, and could leak behavior from the writing engine to the reading engine. For example, double-to-string casts can produce different results in different engines. Performing such a cast while shredding (even if we somehow retained the knowledge that the original value was a `double`) could result in confusing behavior changes if shredding took place using a different tool from the query engine that wrote it.
+
+Our approach is a pragmatic compromise that allows the use of **typed_value** in cases where the type can be losslessly widened without resulting in a significant difference in the reconstructed Variant:
```

Comment:
It seems an alternative here could be to have multiple typed values "typed_value_int32" "typed_value_int64", if necessary.  This carries additional overhead, was that overhead deemed unacceptable or too complex?

Reply:
A case we were concerned about is ingesting JSON data, which is expected to be a very common use of Variant. In that situation, there can easily be a variety of decimal precisions and integer sizes for a given field. Adding a separate column for each type has two problems that I can see:
1) It may add many sparse columns to the parquet schema, which can have performance implications.
2) Any flow that tries to infer an appropriate parquet schema, without exhaustively sampling the entire file contents could easily miss a particular type, which would then need to be written to the untyped_value, making stats unusable for that column.

I think multiple typed values could be reasonable in some scenarios, but it would be good to clarify the use cases, and be confident that the benefit justifies the extra complexity.

Reply:
The main think I was considering was preserving numeric types unambiguously but still have them shredded.  After sleeping on it a better solution could be having an "original type column" which stores the type ID of the original numeric type that was shredded.  This could be optional if all numeric types were of the same

Reply:
Agreed, I think this is my preferred solution to resolving the concerns that you and @rdblue raised.